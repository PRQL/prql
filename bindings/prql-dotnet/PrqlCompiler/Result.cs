using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;

namespace Prql.Compiler
{
    /// <summary>
    /// Result of compilation.
    /// </summary>
    public class Result
    {
        private readonly IReadOnlyCollection<Message> _messages;

        internal Result(NativeResult result)
        {
            Output = result.Output;

            var messages = new List<Message>();

            for (var i = 0; i < result.MessagesLen; i++)
            {
                messages.Add(Marshal.PtrToStructure<Message>(result.Messages));
            }

            _messages = messages.ToList().AsReadOnly();
        }

        /// <summary>
        /// The compiler output.
        /// </summary>
        public string Output { get; }

        /// <summary>
        /// Error, warning and lint messages.
        /// </summary>
        public IReadOnlyCollection<Message> Messages => _messages;
    }
}
