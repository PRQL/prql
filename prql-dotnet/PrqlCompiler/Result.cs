using System.Collections.Generic;
using System.Linq;

namespace Prql.Compiler
{
    /// <summary>
    /// Result of compilation.
    /// </summary>
    public class Result
    {
        private readonly IReadOnlyCollection<Message> _messages;

        internal Result(string output, IEnumerable<Message> messages)
        {
            Output = output;
            _messages = messages.ToList().AsReadOnly();
        }

        /// <summary>
        /// @var string
        /// </summary>
        public string Output { get; }

        /// <summary>
        /// @var array<Message>
        /// </summary>
        public IReadOnlyCollection<Message> Messages => _messages;
    }
}