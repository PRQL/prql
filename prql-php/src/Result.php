<?php

declare(strict_types=1);

namespace Prql\Compiler;

/**
 * Result of compilation.
 */
final class Result
{
    /**
     * @var string
     */
    public string $output;

    /**
     * @var array<Message>
     */
    public array $messages;
}
