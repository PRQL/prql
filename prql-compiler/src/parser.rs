use std::collections::HashMap;

use anyhow::Result;

use crate::{ast::pl::Stmt, utils::IdGenerator, SourceTree};

mod prql_parser;

pub fn parse(file_tree: &SourceTree<String>) -> Result<SourceTree<Vec<Stmt>>> {
    let mut res = SourceTree::default();

    let ids: HashMap<_, _> = file_tree.source_ids.iter().map(|(a, b)| (b, a)).collect();
    let mut id_gen = IdGenerator::<usize>::new();

    for (path, source) in &file_tree.sources {
        let id = ids
            .get(path)
            .map(|x| **x)
            .unwrap_or_else(|| id_gen.gen() as u16);
        let stmts = parse_source(source, id)?;

        res.sources.insert(path.clone(), stmts);
        res.source_ids.insert(id, path.clone());
    }
    Ok(res)
}

fn parse_source(source: &str, source_id: u16) -> Result<Vec<Stmt>> {
    prql_parser::parse_source(source, source_id)
}
