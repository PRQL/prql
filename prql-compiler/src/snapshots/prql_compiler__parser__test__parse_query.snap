---
source: prql-compiler/src/parser.rs
expression: "ast_of_string(r#\"\nfrom employees\nfilter country == \"USA\"                        # Each line transforms the previous result.\nderive [                                      # This adds columns / variables.\n  gross_salary = salary + payroll_tax,\n  gross_cost =   gross_salary + benefits_cost # Variables can use other variables.\n]\nfilter gross_cost > 0\ngroup [title, country] (\naggregate [               # `by` are the columns to group by.\n                   average salary,            # These are aggregation calcs run on each group.\n                   sum salary,\n                   average gross_salary,\n                   sum gross_salary,\n                   average gross_cost,\n  sum_gross_cost = sum gross_cost,\n  ct             = count,\n] )\nsort sum_gross_cost\nfilter ct > 200\ntake 20\n    \"#.trim(),\n              Rule::query)?"
---
Query:
  version: ~
  dialect: Generic
  nodes:
    - Pipeline:
        nodes:
          - FuncCall:
              name: from
              args:
                - Ident: employees
              named_args: {}
          - FuncCall:
              name: filter
              args:
                - Binary:
                    left:
                      Ident: country
                    op: Eq
                    right:
                      Literal:
                        String: USA
              named_args: {}
          - FuncCall:
              name: derive
              args:
                - List:
                    - Assign:
                        name: gross_salary
                        expr:
                          Binary:
                            left:
                              Ident: salary
                            op: Add
                            right:
                              Ident: payroll_tax
                    - Assign:
                        name: gross_cost
                        expr:
                          Binary:
                            left:
                              Ident: gross_salary
                            op: Add
                            right:
                              Ident: benefits_cost
              named_args: {}
          - FuncCall:
              name: filter
              args:
                - Binary:
                    left:
                      Ident: gross_cost
                    op: Gt
                    right:
                      Literal:
                        Integer: 0
              named_args: {}
          - FuncCall:
              name: group
              args:
                - List:
                    - Ident: title
                    - Ident: country
                - FuncCall:
                    name: aggregate
                    args:
                      - List:
                          - FuncCall:
                              name: average
                              args:
                                - Ident: salary
                              named_args: {}
                          - FuncCall:
                              name: sum
                              args:
                                - Ident: salary
                              named_args: {}
                          - FuncCall:
                              name: average
                              args:
                                - Ident: gross_salary
                              named_args: {}
                          - FuncCall:
                              name: sum
                              args:
                                - Ident: gross_salary
                              named_args: {}
                          - FuncCall:
                              name: average
                              args:
                                - Ident: gross_cost
                              named_args: {}
                          - Assign:
                              name: sum_gross_cost
                              expr:
                                FuncCall:
                                  name: sum
                                  args:
                                    - Ident: gross_cost
                                  named_args: {}
                          - Assign:
                              name: ct
                              expr:
                                Ident: count
                    named_args: {}
              named_args: {}
          - FuncCall:
              name: sort
              args:
                - Ident: sum_gross_cost
              named_args: {}
          - FuncCall:
              name: filter
              args:
                - Binary:
                    left:
                      Ident: ct
                    op: Gt
                    right:
                      Literal:
                        Integer: 200
              named_args: {}
          - FuncCall:
              name: take
              args:
                - Literal:
                    Integer: 20
              named_args: {}

